File: apps/clojure/test/fizzbuzz/application/command_test.clj
(ns fizzbuzz.application.command-test
  (:require [clojure.test :refer :all]
            [fizzbuzz.application.command :as cmd]))

(deftest fizz-buzz-value-command-test
  (testing "FizzBuzzValueCommand: 単一値コマンド"
    (testing "タイプ 1"
      (is (= "1" (cmd/execute (cmd/->FizzBuzzValueCommand 1 1))))
      (is (= "Fizz" (cmd/execute (cmd/->FizzBuzzValueCommand 3 1))))
      (is (= "Buzz" (cmd/execute (cmd/->FizzBuzzValueCommand 5 1))))
      (is (= "FizzBuzz" (cmd/execute (cmd/->FizzBuzzValueCommand 15 1)))))
    (testing "タイプ 2"
      (is (= "1" (cmd/execute (cmd/->FizzBuzzValueCommand 1 2))))
      (is (= "3" (cmd/execute (cmd/->FizzBuzzValueCommand 3 2)))))
    (testing "タイプ 3"
      (is (= "" (cmd/execute (cmd/->FizzBuzzValueCommand 1 3))))
      (is (= "Fizz" (cmd/execute (cmd/->FizzBuzzValueCommand 3 3)))))))

(deftest fizz-buzz-list-command-test
  (testing "FizzBuzzListCommand: リストコマンド"
    (let [result (cmd/execute (cmd/->FizzBuzzListCommand (range 1 101) 1))]
      (is (= 100 (count (:values result))))
      (is (= "1" (first (:values result))))
      (is (= "Buzz" (last (:values result)))))))

(deftest fizz-buzz-command-error-test
  (testing "未定義タイプでコマンド実行すると例外が発生する"
    (is (thrown? clojure.lang.ExceptionInfo
                 (cmd/execute (cmd/->FizzBuzzValueCommand 1 99))))))

File: apps/elixir/test/fizz_buzz/command_test.exs
defmodule FizzBuzz.CommandTest do
  use ExUnit.Case

  alias FizzBuzz.Command.{ValueCommand, ListCommand}
  alias FizzBuzz.Type.Type01
  alias FizzBuzz.Model

  describe "ValueCommand" do
    test "デフォルトで実行できる" do
      result = ValueCommand.execute()
      assert %Model.Value{} = result
    end

    test "指定した数値で実行できる" do
      cmd = %ValueCommand{number: 3, type: %Type01{}}
      result = ValueCommand.execute(cmd)
      assert result.value == "Fizz"
    end

    test "15 の倍数で FizzBuzz を返す" do
      cmd = %ValueCommand{number: 15, type: %Type01{}}
      result = ValueCommand.execute(cmd)
      assert result.value == "FizzBuzz"
    end
  end

  describe "ListCommand" do
    test "デフォルトで 100 件のリストを生成する" do
      result = ListCommand.execute()
      assert Model.List.count(result) == 100
    end

    test "指定した件数でリストを生成する" do
      cmd = %ListCommand{count: 10, type: %Type01{}}
      result = ListCommand.execute(cmd)
      assert Model.List.count(result) == 10
    end
  end
end
